control-plane-operator/controllers/hostedcontrolplane/etcddefrag/deployment.go

```diff
@@ -0,0 +1,131 @@
+package etcddefrag
+
+import (
+	"context"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1beta1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/manifests"
+	"github.com/openshift/hypershift/support/config"
+	"github.com/openshift/hypershift/support/util"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+const (
+	configHashAnnotation = "etcddefrag.hypershift.openshift.io/config-hash"
+)
+
+var (
+	volumeMounts = util.PodVolumeMounts{
+		etcdDefragContainerMain().Name: {
+			etcdDefragVolumeClientCert().Name: "/var/run/secrets/etcd-client",
+			etcdDefragVolumeTrustedCA().Name:  "/var/run/configmaps/etcd-ca/ca.crt",
+		},
+	}
+)
+
+func etcdDefragLabels() map[string]string {
+	return map[string]string{
+		"app":                         "etcd-defrag-controller-openshift",
```

Comment:
I see this pull adding a defrag controller, but [HOSTEDCP-1122](https://issues.redhat.com/browse/HOSTEDCP-1122) lists as a goal:

> etcd defragmentation and compaction should be performed as needed for HA deployments

and I'm not seeing anything around [compaction][1] ([which is handled by the Kubernetes API server in standalone OpenShift][2], although I haven't been able to track down the backing configuration).  Am I missing something in the pull?  Is that no longer in-scope?  Is it still in-scope for the epic but being moved to a follow-up pull request?

[1]: https://etcd.io/docs/v3.5/op-guide/maintenance/#auto-compaction
[2]: https://github.com/openshift/cluster-etcd-operator/blob/2e1bb31d9d0a9725d8ef0604580a3e70ceeff0a5/docs/performance-metrics.md?plain=1#L32

Reply:
[Recent 4.14 nightly][1] -> [aggregated-hypershift-ovn-conformance][2] -> [run][3]:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-hypershift-release-4.14-periodics-e2e-aws-ovn-conformance/1706651913035976704/artifacts/e2e-aws-ovn-conformance/dump/artifacts/namespaces/clusters-3557d0e7d5a547dbdb99/core/pods/logs/kube-apiserver-58784597c5-4jx2d-kube-apiserver.log | grep compact
I0926 13:17:01.488231       1 flags.go:64] FLAG: --etcd-compaction-interval="5m0s"
I0926 13:17:01.490281       1 flags.go:64] FLAG: --etcd-compaction-interval="5m0s"
W0926 13:28:24.404804       1 watcher.go:249] watch chan error: etcdserver: mvcc: required revision has been compacted
W0926 13:37:44.450747       1 watcher.go:249] watch chan error: etcdserver: mvcc: required revision has been compacted
W0926 13:40:24.686254       1 watcher.go:249] watch chan error: etcdserver: mvcc: required revision has been compacted
W0926 13:51:30.915717       1 watcher.go:249] watch chan error: etcdserver: mvcc: required revision has been compacted
...
```

And kubernetes/kubernetes#51765 includes:

> ... If the flag is true (default), apiserver issues the compaction requests like current behaviour...

so  maybe upstream has us covered on the compaction side of things without us needing to do anything, and the Jira tickets can be scoped down to just talking about defrag?

[1]: https://amd64.ocp.releases.ci.openshift.org/releasestream/4.14.0-0.nightly/release/4.14.0-0.nightly-2023-09-26-124507
[2]: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/aggregated-hypershift-ovn-conformance-release-openshift-release-analysis-aggregator/1706651924629032960
[3]: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-hypershift-release-4.14-periodics-e2e-aws-ovn-conformance/1706651913035976704

Reply:
Yes it also does compaction.